activiti-cloud-service-common/activiti-cloud-services-test/src/main/java/org/activiti/cloud/starters/test/EventsAggregator.java

```diff
@@ -18,24 +18,59 @@
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
 import org.activiti.cloud.api.model.shared.events.CloudRuntimeEvent;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageHandler;
+import org.springframework.messaging.MessagingException;
+import org.springframework.messaging.SubscribableChannel;
+import org.springframework.messaging.support.ErrorMessage;
 
-public class EventsAggregator {
+public class EventsAggregator implements MessageHandler {
 
     private List<CloudRuntimeEvent<?, ?>> events = new ArrayList<>();
 
     private MyProducer producer;
 
+    private AtomicReference<Message<?>> errorMessageRef = new AtomicReference<>();
+
     public EventsAggregator(MyProducer producer) {
         this.producer = producer;
     }
 
-    public void addEvents(CloudRuntimeEvent<?, ?>... events) {
+    public EventsAggregator addEvents(CloudRuntimeEvent<?, ?>... events) {
         this.events.addAll(Arrays.asList(events));
+
+        return this;
     }
 
-    public void sendAll() {
+    public CloudRuntimeEvent<?, ?>[] sendAll() {
+        List<CloudRuntimeEvent<?, ?>> sentEvents = new ArrayList<>(events);
+
+        errorMessageRef.set(null);
         producer.send(events.toArray(new CloudRuntimeEvent<?, ?>[] {}));
         events.clear();
+
+        return sentEvents.toArray(CloudRuntimeEvent[]::new);
+    }
+
+    @Override
+    public void handleMessage(Message<?> message) throws MessagingException {
+        errorMessageRef.set(message);
+    }
+
+    public Throwable getException() {
+        return Optional
+            .ofNullable(errorMessageRef.get())
+            .map(ErrorMessage.class::cast)
+            .map(ErrorMessage::getPayload)
+            .orElse(null);
+    }
+
+    public EventsAggregator errorChannel(SubscribableChannel errorChannel) {
+        errorChannel.subscribe(this);
+
+        return this;
```

Comment:
Is this method and all the references to `errorMessageRef` used by tests only? If that's the case, wouldn't it be better to create a dedicated `MessageHandler` in the test scope only?

Reply:
It is a test component used always in the test scope. We can refactor this in the next iteration.